### PR TITLE
Use filter_map instead of filter() and map()

### LINKED ISSUE #5436: Use`filter_map()` instead of `filter` and then a`map`
I think it may be more efficient to use a `filter_map()` here, no?

_Originally posted by @Braqzen in https://github.com/FuelLabs/sway/pull/5435#discussion_r1442317525_

### PR DESCRIPTION
**Pull request by crodas**
Fixes #5436

## Description

Replace chained `filter` and then `map` calls for a single `filter_map` calls and other minor optimizations

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: sway-lsp/src/utils/debug.rs [hunk 1]
```diff
@@ -11,8 +11,13 @@ where
     I: Iterator<Item = (TokenIdent, Token)>,
 {
     tokens
-        .filter(|(_, token)| token.typed.is_none())
-        .map(|(ident, _)| warning_from_ident(&ident))
+        .filter_map(|(ident, token)| {
+            if token.typed.is_none() {
```
Comment (sdankel) [inline sway-lsp/src/utils/debug.rs:15]: nit: this could be more readable with a `match`

File: sway-lsp/src/utils/debug.rs [hunk 2]
```diff
@@ -30,8 +35,13 @@ where
     I: Iterator<Item = (TokenIdent, Token)>,
 {
     tokens
-        .filter(|(_, token)| token.typed.is_some())
-        .map(|(ident, _)| warning_from_ident(&ident))
+        .filter_map(|(ident, token)| {
+            if token.typed.is_some() {
```
Comment (sdankel) [inline sway-lsp/src/utils/debug.rs:39]: nit: this could be more readable with a `match`

**Review by sdankel — APPROVED** [review]
LGTM

**Review by kayagokalp — COMMENTED** [review]
Changes looks good to me but I am a little lost about the reason we need this. As far as I am aware of, `filter_map` and `filter` then `map` is not the same in terms of produced code by the rust compiler and I feel like it won't be the case that `filter_map` is always faster. But I haven't read the std code for this and not sure about the above statement, happy to approve

File: forc-pkg/src/lock.rs [hunk 3]
```diff
@@ -90,13 +90,23 @@ impl PkgLock {
...
         let mut dependencies: Vec<String> = all_dependencies
             .iter()
-            .filter(|(_, dep_kind)| *dep_kind == DepKind::Library)
-            .map(|(dep_pkg, _)| dep_pkg.clone())
+            .filter_map(|(dep_pkg, dep_kind)| {
+                if *dep_kind == DepKind::Library {
+                    Some(dep_pkg.clone())
+                } else {
+                    None
+                }
```
Comment (IGI-111) [inline forc-pkg/src/lock.rs:98]:
```suggestion
                (*dep_kind == DepKind::Library).then_some(dep_pkg.clone())
```

File: forc-pkg/src/lock.rs [hunk 4]
```diff
@@ -90,13 +90,23 @@ impl PkgLock {
...
         let mut contract_dependencies: Vec<String> = all_dependencies
             .iter()
-            .filter(|(_, dep_kind)| matches!(*dep_kind, DepKind::Contract { .. }))
-            .map(|(dep_pkg, _)| dep_pkg.clone())
+            .filter_map(|(dep_pkg, dep_kind)| {
+                if matches!(*dep_kind, DepKind::Contract { .. }) {
+                    Some(dep_pkg.clone())
+                } else {
+                    None
+                }
```
Comment (IGI-111) [inline forc-pkg/src/lock.rs:108]:
```suggestion
                matches!(*dep_kind, DepKind::Contract { .. }).then_some(dep_pkg.clone())
```

File: forc-pkg/src/manifest.rs [hunk 5]
```diff
@@ -1040,8 +1043,14 @@ impl std::ops::Deref for WorkspaceManifestFile {
...
         .into_iter()
-        .filter_map(Result::ok)
-        .filter(|entry| entry.path().ends_with(constants::MANIFEST_FILE_NAME))
+        .filter_map(|entry| {
+            if let Ok(entry) = entry {
+                if entry.path().ends_with(constants::MANIFEST_FILE_NAME) {
+                    return Some(entry);
+                }
+            }
+            None
```
Comment (IGI-111) [inline forc-pkg/src/manifest.rs:1052]:
```suggestion
            entry
                .ok()
                .filter(|entry| entry.path().ends_with(constants::MANIFEST_FILE_NAME))
```

File: forc-plugins/forc-doc/src/doc/mod.rs [hunk 6]
```diff
@@ -60,12 +60,15 @@ impl Documentation {
...
-                )
+            .filter_map(|d| {
+                if d.item_header.friendly_name == "trait" {
+                    Some((
+                        d.item_header.item_name.clone(),
+                        d.item_header.module_info.clone(),
+                    ))
+                } else {
+                    None
+                }
```
Comment (IGI-111) [inline forc-plugins/forc-doc/src/doc/mod.rs:71]:
```suggestion
                (d.item_header.friendly_name == "trait").then_some((
                    d.item_header.item_name.clone(),
                    d.item_header.module_info.clone(),
                ))
```

File: forc/src/cli/plugin.rs [hunk 7]
```diff
@@ -93,6 +93,12 @@ pub(crate) fn find_all() -> impl Iterator<Item = PathBuf> {
...
         .filter_map(Result::ok)
-        .map(|entry| entry.path().to_path_buf())
-        .filter(|p| is_plugin(p))
+        .filter_map(|entry| {
+            let path = entry.path().to_path_buf();
+            if is_plugin(&path) {
+                Some(path)
+            } else {
+                None
+            }
```
Comment (IGI-111) [inline forc/src/cli/plugin.rs:102]:
```suggestion
            is_plugin(&path).then_some(path)
```

File: sway-core/src/asm_generation/fuel/optimizations.rs [hunk 8]
```diff
@@ -234,8 +234,13 @@ impl AbstractInstructionSet {
...
             .into_iter()
             .enumerate()
-            .filter(|(idx, _)| !dead_indices.contains(idx))
-            .map(|(_, op)| op)
+            .filter_map(|(idx, op)| {
+                if !dead_indices.contains(&idx) {
+                    Some(op)
+                } else {
+                    None
+                }
```
Comment (IGI-111) [inline sway-core/src/asm_generation/fuel/optimizations.rs:242]:
```suggestion
                !dead_indices.contains(&idx).then_some(op)
```

File: sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs [hunk 9]
```diff
@@ -299,8 +299,13 @@ fn match_or(
...
             let missing_in_alternatives: Vec<Span> = variables_in_alternatives
                 .iter()
-                .filter(|(_, vars)| !vars.iter().any(|(ident, _)| ident == *variable))
-                .map(|(span, _)| span.clone())
+                .filter_map(|(span, vars)| {
+                    if !vars.iter().any(|(ident, _)| ident == *variable) {
+                        Some(span.clone())
+                    } else {
+                        None
+                    }
```
Comment (IGI-111) [inline sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs:307]:
```suggestion
                    (!vars.iter().any(|(ident, _)| ident == *variable)).then_some(span.clone())
```

File: sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs [hunk 10]
```diff
@@ -324,11 +329,13 @@ fn match_or(
...
-                    .filter(|(ident, var_type_id)| {
-                        ident == variable && !equality.check(type_id, *var_type_id)
-                    })
-                    .map(|(ident, var_type_id)| (ident.clone(), var_type_id))
+                vars.iter().filter_map(|(ident, var_type_id)| {
+                    if ident == variable && !equality.check(type_id, *var_type_id) {
+                        Some((ident.clone(), *var_type_id))
+                    } else {
+                        None
+                    }
```
Comment (IGI-111) [inline sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs:337]:
```suggestion
                    (ident == variable && !equality.check(type_id, *var_type_id)).then_some((ident.clone(), *var_type_id))
```

File: sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/typed_scrutinee.rs [hunk 11]
```diff
@@ -270,8 +270,13 @@ fn type_check_struct(
...
             .fields
             .iter()
-            .filter(|f| !typed_fields.iter().any(|tf| f.name == tf.field))
-            .map(|f| f.name.to_string())
+            .filter_map(|f| {
+                if !typed_fields.iter().any(|tf| f.name == tf.field) {
+                    Some(f.name.to_string())
+                } else {
+                    None
+                }
```
Comment (IGI-111) [inline sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/typed_scrutinee.rs:278]:
```suggestion
                (!typed_fields.iter().any(|tf| f.name == tf.field)).then_some(f.name.to_string())
```

File: sway-ir/src/optimize/mem2reg.rs [hunk 12]
```diff
@@ -48,13 +48,17 @@ fn filter_usable_locals(context: &mut Context, function: &Function) -> HashSet<S
...
             let ty = var.get_inner_type(context);
-            ty.is_unit(context)
+            if ty.is_unit(context)
                 || ty.is_bool(context)
                 || (ty.is_uint(context) && ty.get_uint_width(context).unwrap() <= 64)
+            {
+                Some(name.clone())
+            } else {
+                None
+            }
```
Comment (IGI-111) [inline sway-ir/src/optimize/mem2reg.rs:60]:
```suggestion
            (ty.is_unit(context)
                || ty.is_bool(context)
                || (ty.is_uint(context) && ty.get_uint_width(context).unwrap() <= 64))
                .then_some(name.clone())
```

File: sway-lsp/src/utils/debug.rs [hunk 13]
```diff
@@ -11,8 +11,13 @@ where
...
     tokens
-        .filter(|(_, token)| token.typed.is_none())
-        .map(|(ident, _)| warning_from_ident(&ident))
+        .filter_map(|(ident, token)| {
+            if token.typed.is_none() {
+                Some(warning_from_ident(&ident))
+            } else {
+                None
+            }
+        })
```
Comment (IGI-111) [inline sway-lsp/src/utils/debug.rs:20]:
```suggestion
        .filter_map(|(ident, token)| token.typed.is_none().then_some(warning_from_ident(&ident)))

```

File: sway-lsp/src/utils/markdown.rs [hunk 14]
```diff
@@ -53,8 +53,14 @@ fn is_sway_fence(s: &str) -> bool {
...
         .split(|c| matches!(c, ',' | ' ' | '\t'))
-        .map(str::trim)
-        .filter(|t| !t.is_empty());
+        .filter_map(|t| {
+            let t = t.trim();
+            if t.is_empty() {
+                None
+            } else {
+                Some(t)
+            }
```
Comment (IGI-111) [inline sway-lsp/src/utils/markdown.rs:62]:
```suggestion
            (!t.is_empty()).then_some(t)
```

**Review by JoshuaBatty — APPROVED** [review]

**Review by IGI-111 — APPROVED** [review]

**Pull request merged**